### ISSUE TITLE
feat: user configurable diagnostic source name

### ISSUE THREAD

**Issue by JA-Bar** [jose-elias-alvarez/null-ls.nvim#520]
### Issues

- [X] I have checked [existing issues](https://github.com/jose-elias-alvarez/null-ls.nvim/issues) and there are no existing ones with the same request.

### Feature description

Ever since 859a3e0b1344e1bb937c1321e491b48dd05a1f24 passing the "name" option to a builtin source no longer changes the diagnostic's source name when displaying diagnostics, so if you have a custom command it looks like this:
```
Diagnostics:
line too long (93 > 90 characters) (/home/user/.local/share/nvim/installables/python/bin/flake8) 
```
instead of:
```
Diagnostics:
line too long (93 > 90 characters) (my_custom_name) 
```

I don't know if this was intended behavior, but it would be nice to have it back as a feature.

### Help

Yes, but I don't know how to start. I would need guidance

### Implementation help

Merging the user's "name" option with the default generator options solves the issue. 
```diff
diff --git a/lua/null-ls/helpers/make_builtin.lua b/lua/null-ls/helpers/make_builtin.lua
index 0265014..a13db7d 100644
--- a/lua/null-ls/helpers/make_builtin.lua
+++ b/lua/null-ls/helpers/make_builtin.lua
@@ -43,6 +43,7 @@ local function make_builtin(opts)
         cwd = opts.cwd,
         diagnostics_format = opts.diagnostics_format,
         dynamic_command = opts.dynamic_command,
+        name = opts.name,
         runtime_condition = opts.runtime_condition,
         timeout = opts.timeout,
     })
```
But fails the test:
```
/home/user/.local/share/nvim/site/pack/packer/start/null-ls.nvim/test/spec/helpers/make_builtin_spec.lua                               
Fail    ||      make_builtin should return builtin with assigned opts                                                                                  
            ...art/null-ls.nvim/test/spec/helpers/make_builtin_spec.lua:36: Expected objects to be the same.
```
So I don't know if this feature was just thrown out or if it was an oversight of the last refactoring.

**Comment by jose-elias-alvarez** [comment]
The fact that this worked before was an unintended side effect of indiscriminately treating all user options as generator options, but it should be fixed by 295c4b790e2a54106dd1f396e1b1dc18c19a6db8. Can you let me know?

Separately, all built-in sources should probably have a `name` field defined so that overriding `command` doesn't result in this behavior. I'll see if I can put together a script.

**Comment by JA-Bar** [comment]
I can confirm that it solves the issue, thanks!

As far as naming the sources goes, I think that should indeed be the case, although I believe currently the source's name has priority over the user-passed "name" option, so if the default source name is "flake8", but I want the diagnostic source to be named "strict-flake8", the source would show as "flake8". This is left as a design decision to you in case you want to fully support the feature.

I can submit a quick PR tomorrow adding names to built-ins if you'd like.

**Comment by jose-elias-alvarez** [comment]
I agree that setting a default name but letting users override it is the best approach. PR greatly appreciated!

**Comment by jose-elias-alvarez** [comment]
Closed by 295c4b790e2a54106dd1f396e1b1dc18c19a6db8 and 031e839ac2a8ca1a4bc4111633a829dcae741c31.
